pyiron_atomistics/sphinx/output_parser.py

```diff
@@ -417,3 +419,95 @@ def results(self):
             if len(value) > 0:
                 results["dft"][key] = value
         return results
+
+
+class SphinxWavesReader:
+    """Class to read SPHInX waves.sxb files (HDF5 format)
+
+    Initialize with waves.sxb filename, or use load ()
+
+    """
+
+    def __init__(self, file_name="waves.sxb", cwd="."):
+        """
+        Args:
+            file_name (str): file name
+            cwd (str): directory path
+        """
+        if Path(file_name).is_absolute():
+            self.wfile = h5py.File(Path(file_name))
+        else:
+            path = Path(cwd) / Path(file_name)
+            self.wfile = h5py.File(path)
+
+    @property
+    def _n_gk(self):
+        return self.wfile['nGk'][:]
+    
+    @cached_property
+    def mesh(self):
+        return self.wfile["meshDim"][:]
+
+    @property
+    def Nx(self):
+        return self.mesh[0]
+
+    @property
+    def Ny(self):
+        return self.mesh[1]
+
+    @property
+    def Nz(self):
+        return self.mesh[2]
+
+    @cached_property
+    def n_states(self):
+        return self.wfile["nPerK"][0]
+
+    @cached_property
+    def n_spin(self):
+        return self.wfile["nSpin"].shape[0]
+
+    @cached_property
+    def k_weights(self):
+        return self.wfile["kWeights"][:]
+
+    @cached_property
+    def k_vec(self):
+        return self.wfile["kVec"][:]
+
+    @cached_property
+    def eps(self):
+        """All eigenvalues (in Hartree) as (nk,n_states) block"""
+        return (self.wfile['eps'][:].reshape(-1, self.n_spin, self.n_states)).T
+
+    @cached_property
+    def _fft_idx(self):
+        fft_idx = []
+        off = 0
+        for ngk in self._n_gk:
+            fft_idx.append(self.wfile["fftIdx"][off : off + ngk])
+            off += ngk
+        return fft_idx
+
+    def get_psi_rec(self, i, ispin, ik):
+        """Loads a single wavefunction on full FFT mesh"""
+        if i < 0 or i >= self.n_states:
+            raise IndexError(f"i={i} fails 0 <= i < n_states={self.n_states}")
```

Comment:
```suggestion
        i = np.arange(self.n_states)[i]
```

Failing `i < 0` is actually a bit against the idea of python, as for example -1 means `self.n_states - 1` in this case. The operation I suggest above will already include the message when `i < n_states`. The same applies in the if statements below.

Reply:
I don't understand. If `i = -1 ` then the code should throw an error as states are indexed from 0 to `self.n_states`. And I would think if the user uses a state index higher than or less than zero, this IndexError makes sense.

Reply:
How is it different from `my_list[index]`? Here, `index` can be also a negative number

Reply:
by index I mean how a state is labeled. this function basically gives a wavefunction corresponding to a state labelled by (state,spin,k).

Reply:
yeah, and -1 means the last state

Reply:
I understand this from python. But coming from SPHInX, I felt all the states are always labelled by positive numbers and negative number should through and error rather than giving last states.

Reply:
I do see your point. But I will wait to see @freyso opinion on this.

Reply:
Sure, because SPHInX is written in C++, which doesn't allow negative indices, but it shouldn't have a fundamental problem with allowing it. Not allowing negative indices however breaks the python convention, and I don't think it's a good idea to do so only because we don't know what it could be useful for. If there's a fundamental problem, I'm also ok to forbid it.

Reply:
I agree to the suggested workaround to ensure that negative indices are handled in a pythonic way. Just a comment in the code would be good, something like "translate indices from pythonic style to 0 ... n-1". Otherwise, the lines look like a complicated idempotency.